### PR TITLE
Porting rosapi package.

### PR DESCRIPTION
**Pull request by jubeira**
This PR focuses on porting `rosapi` package to ROS2.

See https://github.com/ros2/rosbridge_suite/issues/2 for the current overall progress.

### CODE REVIEW

**Comment by jubeira** [comment]
@dirk-thomas this should be ready for an initial review. Some notes to consider:
- There are two services that haven't been ported: `ServiceHost` and `SearchParam`. 
  - `ServiceHost` should return the hostname of the node providing a given service. This is not available in `rclpy`. For topics for example, only the [amount of publishers / subscribers](https://github.com/ros2/ros2cli/pull/88) can be retrieved in contrast with `rostopic info` in ROS1, where the hostname is also exposed. I've taken a quick look and adding this would mean adjusting the code all the way down to `rmw` layer to upstream that information.
  - `SearchParam` relied on `rospy.search_param`. Porting it would require bringing in that functionality to ROS2 and maybe adapting it a bit. It shouldn't be as hard as `ServiceHost`; it's just not straightforward and I'm not sure whether this is needed or not for a start.
- For services related with parameters:
  - I created a [different node](https://github.com/RobotWebTools/rosbridge_suite/pull/388/files#diff-9b18b072495ba40106802093bb486fb5R62) to call services within service callbacks without entering a deadlock. There's a note explaining that it should be a temporary workaround.
  - I added `node_name` as a parameter for service requests, as parameters are now associated with nodes in contrast to ROS1. This breaks the API, but for a start it seemed the most logical thing to do. Another possibility would be to guess the node name and the parameter name from a single string considering the node names can be fetched before, but that is a bit more complex.
- Adding test cases would be a good idea.

That being said, my initial suggestion would be to proceed if possible without dealing with those two remaining services. Of course, comments and suggestions are welcome!

File: rosbridge_library/src/rosbridge_library/internal/ros_loader.py [hunk 1]
```diff
@@ -177,26 +179,46 @@ def _load_class(modname, subname, classname):
...
+
+    Updates system path to include the path of a given package.
+    """
+    global _loaded_modules
+
+    if modname in _loaded_modules:
+        return
+
+    prefix_path = get_prefix_path(modname)
+    package_path = os.path.join(prefix_path, 'lib', 'python3.{}'.format(sys.version_info.minor), 'site-packages')
```
Comment (dirk-thomas) [inline rosbridge_library/src/rosbridge_library/internal/ros_loader.py]: This logic is not portable. What is this needed for? All Python packages should already be in the path.

**Comment by dirk-thomas** [comment]
> There are two services that haven't been ported:

What are those two services being used? So how severe is it to not port them?

**Comment by jubeira** [comment]
Thanks for the comments @dirk-thomas! I'll be addressing them soon.

> What are those two services being used? So how severe is it to not port them?

These services aren't being used in rosbridge itself; I've searched the codebase for a client for those services and I found none.
I also made some tests with `rvizweb` by adding some print statements for each service handler and then using the controls provided by `rvizweb` (pointcloud, image, depthcloud, tf, etc). The only service that was used was `GetParam`, the others were not.

In other words, a missing service might affect client code using `rosbridge`, not `rosbridge` itself. Not porting those two services in particular doesn't seem to have a very high impact at least in principle.

File: rosbridge_library/src/rosbridge_library/internal/ros_loader.py (hunk 1 above)
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/ros_loader.py]: The original version of `rosbridge_library` makes use of `roslib` to load packages that might not be in the path, so I wanted to have a minimal version of that functionality so that the `ros2` version works as the `ros1` version as much as possible.

If it's safe to assume that all Python packages are in the path, then this code can be removed and we can just import the required packages directly. I'll do some tests just to double check.

File: rosapi/src/rosapi/objectutils.py [hunk 2]
```diff
@@ -95,7 +95,7 @@ def get_service_response_typedef_recursive(servicetype):
 
 def _get_typedef(instance):
     """ Gets a typedef dict for the specified instance """
-    if instance is None or not hasattr(instance, "__slots__") or not hasattr(instance, "_slot_types"):
+    if instance is None or not hasattr(instance, "__slots__") or not hasattr(instance, "_fields_and_field_types"):
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/objectutils.py]: Thanks for the heads up; I'll keep an eye on that PR and update this code once it's merged.

**Comment by jubeira** [comment]
@dirk-thomas comments were addressed; [this one](https://github.com/RobotWebTools/rosbridge_suite/pull/388#discussion_r267036081) is pending until the referenced PR is merged.

File: rosbridge_library/src/rosbridge_library/internal/ros_loader.py (hunk 1 above)
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/ros_loader.py]: I ended up removing this part; as you indicated above it should be safe to assume that the packages are in the path already.

File: rosapi/scripts/rosapi_node [hunk 3]
```diff
@@ -31,150 +31,190 @@
...
+        self.create_service(MessageDetails, '/rosapi/message_details', self.get_message_details)
+        self.create_service(ServiceRequestDetails, '/rosapi/service_request_details', self.get_service_request_details)
+        self.create_service(SetParam, '/rosapi/set_param', self.set_param)
+        self.create_service(GetParam, '/rosapi/get_param', self.get_param)
+        self.create_service(HasParam, '/rosapi/has_param', self.has_param)
+        self.create_service(DeleteParam, '/rosapi/delete_param', self.delete_param)
+        self.create_service(GetParamNames, '/rosapi/get_param_names', self.get_param_names)
+        self.create_service(GetTime, '/rosapi/get_time', self.get_time)
+
+        # Services to be implemented:
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/scripts/rosapi_node]: Addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/9665be8be322ebd895abd6fafbf7a889d6685f2e.

File: rosapi/src/rosapi/glob_helper.py [hunk 4]
```diff
@@ -1,36 +1,36 @@
...
 
-import rospy
+from collections import namedtuple
 import fnmatch
+from rcl_interfaces.msg import ParameterType
 
-topics_glob = []
-services_glob = []
-params_glob = []
+Globs = namedtuple('Globs', ['topics_glob', 'services_glob', 'params_glob'])
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/glob_helper.py]: Addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/9665be8be322ebd895abd6fafbf7a889d6685f2e.

File: rosapi/src/rosapi/glob_helper.py [hunk 5]
```diff
@@ -1,36 +1,36 @@
...
-    global services_glob
-    global params_glob
-
-    def get_param(param_name):
-        param = rospy.get_param(param_name, '')
+    def get_param(parameter_name):
+        parameter_value = node.get_parameter(parameter_name).get_parameter_value()
+        if parameter_value.type == ParameterType.PARAMETER_NOT_SET:
+            parameter_value = ''
+        else:
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/glob_helper.py]: Addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/9665be8be322ebd895abd6fafbf7a889d6685f2e.

File: rosapi/src/rosapi/params.py [hunk 6]
```diff
@@ -55,48 +76,140 @@ def set_param(name, value, params_glob):
...
+    else:
+        parameter.value = ParameterValue()
+        parameter.value.type = parameter_type
+        if parameter_type != ParameterType.PARAMETER_NOT_SET:
+            setattr(parameter.value, _parameter_type_mapping[parameter_type])
+
+    try:
+        # call_get_parameters will fail if node does not exist.
+        call_set_parameters(node=_node, node_name=node_name, parameters=[parameter])
+    except:
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/params.py]: Fixed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/a5b4fecf0ab4f1d570908d143380bdce8793b3b3.

File: rosapi/src/rosapi/params.py [hunk 7]
```diff
@@ -55,48 +76,140 @@ def set_param(name, value, params_glob):
...
+def get_param_names(node_name, params_glob):
+    """ Gets list of parameter names for a given node """
+
+    node_name = get_absolute_node_name(node_name)
+
+    with param_server_lock:
+        if params_glob:
+            # If there is a parameter glob, filter by it.
+            return list(
+                   filter(
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/params.py]: Fixed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/9cebbc7345e16d86799984984db6ac3eb88288dc.

File: rosapi/src/rosapi/params.py [hunk 8]
```diff
@@ -55,48 +76,140 @@ def set_param(name, value, params_glob):
...
     # If the glob list is empty (i.e. false) or the parameter matches
     # one of the glob strings, continue to get the parameter.
-    d = None
     if default is not "":
         try:
-            d = loads(default)
+            default = loads(default)
         except ValueError:
-            d = default
+            pass    # Keep default without modifications.
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/params.py]: Addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/9cebbc7345e16d86799984984db6ac3eb88288dc.

File: rosapi/src/rosapi/objectutils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/objectutils.py]: It seems that in the end it was not replaced (see [rosidl_generator_py/_msg.py.em](https://github.com/ros2/rosidl_python/blob/master/rosidl_generator_py/resource/_msg.py.em)).
It should then be fine to keep this as is.

File: rosapi/src/rosapi/objectutils.py [hunk 9]
```diff
@@ -108,7 +108,7 @@ def _get_typedef(instance):
         fieldnames.append(name)
 
         # Pull out the type and determine whether it's an array
-        field_type = instance._slot_types[i]
+        field_type = instance._fields_and_field_types[name[1:]] # Remove trailing underscore.
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosapi/src/rosapi/objectutils.py]: Addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/388/commits/9cebbc7345e16d86799984984db6ac3eb88288dc.

**Comment by jubeira** [comment]
@dirk-thomas I went through the comments once again to double check; I think this is ready to be merged to `ros2` branch already.
I'll be waiting for approval or further comments. Thanks in advance!

File: rosapi/src/rosapi/params.py [hunk 10]
```diff
@@ -55,48 +76,139 @@ def set_param(name, value, params_glob):
...
     except ValueError:
         raise Exception("Due to the type flexibility of the ROS parameter server, the value argument to set_param must be a JSON-formatted string.")
+
+    node_name = get_absolute_node_name(node_name)
     with param_server_lock:
-        rospy.set_param(name, d)
-    
-    
-def get_param(name, default, params_glob):
+        _set_param(node_name, name, value)
```
Comment (travipross) [inline rosapi/src/rosapi/params.py:82]: @jubeira Sorry to dig up this old PR, but I'm having some trouble with a client that calls on this API and I think I've isolated the issue to this spot. It seems as though when `value` is a JSON-encoded string, the later call to `get_parameter_value()` sets the `ParameterValue.string_value` attribute with double-quotes, meaning it's impossible to set a string parameter via this method without double quotes being applied.

If the client sends the `value` without JSON encoding, the above `ValueError` is thrown and `set_param()` fails altogether.

I think inside the `try` block, there could be a line `value = d if isinstance(d, str) else value`.

This would ensure that a valid string object is later passed to `get_parameter_value()` rather than a JSON-encoded one, but otherwise behaviour would stay the same. Would this potentially work?

Comment (jubeira) [reply rosapi/src/rosapi/params.py:82]: Hi @travipross!
Honestly it's been a while so I'm not sure what the actual consequences of that change could be. Your proposal makes sense but I can't be 100% sure without trying the code.
I'd suggest you to fork the repo, apply the change and check if it works for you. If it does, please feel free to contribute the change with a PR explaining what you did.
Thanks!

Comment (travipross) [reply rosapi/src/rosapi/params.py:82]: Thanks, opened one here: #521
